### PR TITLE
🤖 tests: fuzz gateway, provider routing, and stream error paths; fix error-serialization bugs

### PR DESCRIPTION
**Pull request by ThomasK33**
## Summary

Adds a deterministic fuzzing/chaos pass over the AI Gateway integration, provider routing/resolution, refusal-fallback logic, and stream error handling — and fixes three error-path bugs the campaign surfaced (all in how hostile provider error payloads are stringified/propagated).

## Background

A faulty upstream provider recently hung/broke the agent loop until manual intervention, suggesting hidden edge cases in provider failure handling. This PR probes the mux-owned failure surfaces with seeded fuzz tests: gateway payload normalization, model-string parsing, fallback-chain resolution, route resolution, Coder gateway catalog parsing, error classification, and StreamManager's chunk-consumption loop.

## Implementation

- `src/common/utils/testing/fuzzHelpers.ts`: dependency-free seeded PRNG (mulberry32) plus hostile value/string generators (prototype-pollution keys, lone surrogates, cyclic objects, BigInt, throwing getters/Proxies). Every fuzz test reports its seed, so failures replay deterministically.
- Seven `*.fuzz.test.ts` / chaos suites asserting **totality** (never throw), **idempotence** (normalization survives persist/reload round-trips), **boundedness** (fallback chains ≤ 3, deduped, acyclic; clamped errors stay clamped), and **terminal settlement** (every hostile stream ends in a terminal event with zero unhandled rejections and no wedged workspace; 40 seeded iterations against `StreamManager` with malformed deltas, unknown chunk types, mid-stream throws, truncation, and garbage usage payloads).

Fixes found by the campaign:

1. `getErrorMessage` could throw despite its non-throwing contract: the `String(error)` fallback sat outside the try, so a hostile Proxy / throwing `toString` escaped — in stream-failure paths this masks the real provider error.
2. `getErrorMessage` degraded cyclic/BigInt provider payloads to `"[object Object]"` because bare `JSON.stringify` threw; it now serializes with a cycle/BigInt-safe replacer so the payload's actual fields survive to the UI.
3. `StreamManager`'s error-chunk fallback used bare `JSON.stringify(errorObj)`, replacing cyclic provider errors with a `Converting circular structure to JSON` TypeError; it now routes through `getErrorMessage` (regression test included).

## Validation

- Chaos harness: 40 seeded hostile chunk streams all settle terminally, leak no unhandled rejections, and leave the workspace reusable; a dedicated regression test pins the cyclic error-chunk message behavior.
- Notable non-findings: no hangs/wedges, no unhandled rejections, no prototype-pollution vectors in alias/fallback/route lookups, and gateway model-id mappers roundtrip identity.
- One pre-existing test updated: it pinned the old `"[object Object]"` degradation for circular error objects, which this PR intentionally improves.

## Risks

Low. Production changes are confined to error-message serialization fallbacks (`errors.ts`, one `StreamManager` error-part branch); behavior only changes for error payloads that previously crashed serialization or degraded to `"[object Object]"`. Affected area: error reporting in stream-failure paths. UTF-8/SSE byte-boundary decoding lives inside the AI SDK (not mux-owned), so it is intentionally out of scope; socket-level fault injection against a live gateway would need an integration harness and is noted as follow-up material.

---

_Generated with `xum` • Model: `anthropic:claude-fable-5` • Thinking: `xhigh` • Cost: `$13.41`_

### CODE REVIEW

**Comment by ThomasK33** [comment]
@codex review

**Comment by ThomasK33** [comment]
@codex security review

File: src/node/services/streamManager.chaos.test.ts [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+        // A terminal event must have been emitted to the sink.
+        await new Promise((resolve) => setTimeout(resolve, 0));
+        const terminalTypes = events.filter((event) =>
+          ["stream-end", "error", "stream-abort"].includes(event.type)
+        );
+        expect(terminalTypes.length).toBeGreaterThan(0);
+
+        // The workspace must not be wedged: interruption/cleanup must allow
+        // reuse. stopStream must be a no-op that does not throw.
+        await streamManager.stopStream(workspaceId);
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/node/services/streamManager.chaos.test.ts]: Fixed in 5936eaa65: each chaos iteration now starts a second stream on the same workspace and awaits its settlement, exercising the actual reuse path instead of relying on stopStream's tolerant behavior.

File: src/common/utils/errors.ts [hunk 2]
```diff
@@ -17,7 +17,18 @@ export function getErrorMessage(error: unknown): string {
...
-        const serializedError = JSON.stringify(error);
+        // Cycle/BigInt-safe: provider error payloads can carry circular
+        // references (which make bare JSON.stringify throw) — degrading to
+        // String(error)'s useless "[object Object]".
+        const serializedSeen = new WeakSet<object>();
+        const serializedError = JSON.stringify(error, (_key, value: unknown) => {
+          if (typeof value === "bigint") return value.toString();
+          if (typeof value === "object" && value !== null) {
+            if (serializedSeen.has(value)) return "[Circular]";
+            serializedSeen.add(value);
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/common/utils/errors.ts]: Fixed in 5936eaa65: the serializer now tracks only the current ancestor chain (MDN cycle-detection pattern), so shared sibling references serialize normally and only true cycles become [Circular]. Added a shared-sibling regression test.

File: src/common/constants/coderGateway.fuzz.test.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        // Parsing arbitrary catalog IDs must never throw, and parsed results
+        // must have non-empty origins (empty origins would produce ":model"
+        // pseudo-canonical strings downstream).
+        const junkId = randomFragmentString(rng);
+        if (from) {
+          const parsed = from(junkId);
+          if (parsed) {
+            expect(typeof parsed.origin).toBe("string");
+            expect(typeof parsed.modelId).toBe("string");
+          }
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/common/constants/coderGateway.fuzz.test.ts:88]: Fixed in 5936eaa65: both slash- and dot-separated gateway mappers now reject leading/trailing separators (empty origin or model id), and the fuzz test asserts both parts are non-empty. The strengthened assertion also caught the bedrock dot-mapper's trailing-dot case.

File: src/node/utils/gatewayStreamNormalization.fuzz.test.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      // arithmetic on non-numbers (string concatenation, object coercion).
+      for (const value of [
+        v3.inputTokens.total,
+        v3.inputTokens.noCache,
+        v3.inputTokens.cacheRead,
+        v3.outputTokens.total,
+        v3.outputTokens.text,
+        v3.outputTokens.reasoning,
+      ]) {
+        expect(value === undefined || typeof value === "number").toBe(true);
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/node/utils/gatewayStreamNormalization.fuzz.test.ts]: Fixed in 5936eaa65: flatUsageToV3 now sanitizes non-finite counts (NaN/Infinity) to undefined via a finiteTokenCount guard, and the fuzz test asserts Number.isFinite on every normalized field.

File: src/node/services/streamManager.ts [hunk 5]
```diff
@@ -3629,8 +3629,11 @@ export class StreamManager {
                   errorMessage ??=
                     typeof errorObj.message === "string" ? errorObj.message : undefined;
 
-                  // Last resort: stringify the error
-                  errorMessage ??= JSON.stringify(errorObj);
+                  // Last resort: stringify the error. getErrorMessage guards
+                  // against cyclic/BigInt payloads and undefined-returning
+                  // toJSON, where a bare JSON.stringify would throw and
+                  // replace the provider error with a serialization TypeError.
+                  errorMessage ??= getErrorMessage(errorObj);
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/node/services/streamManager.ts:3639]: Fixed in 5936eaa65: the tool-error path now routes non-Error payloads through getErrorMessage as well, so cyclic/BigInt tool-error payloads stay recoverable tool errors instead of failing the stream.

**Comment by ThomasK33** [comment]
@codex review

**Comment by ThomasK33** [comment]
@codex security review
